core/controllers/cron_test.py

```diff
@@ -0,0 +1,360 @@
+import webtest
+
+(job_models, suggestion_models,) = models.Registry.import_models(
+    [models.NAMES.job, models.NAMES.suggestion])
+
+
+class SampleMapReduceJobManager(jobs.BaseMapReduceJobManager):
+    """Test job that maps over the general suggestion model."""
+
+    @classmethod
+    def entity_classes_to_map_over(cls):
+        return [suggestion_models.GeneralSuggestionModel]
+
+
+class JobStatusMailerHandlerTests(test_utils.GenericTestBase):
+
+    def setUp(self):
+        super(JobStatusMailerHandlerTests, self).setUp()
+        self.signup(self.ADMIN_EMAIL, self.ADMIN_USERNAME)
+        self.admin_id = self.get_user_id_from_email(self.ADMIN_EMAIL)
+        self.set_admins([self.ADMIN_USERNAME])
+        self.testapp_swap = self.swap(
+            self, 'testapp', webtest.TestApp(main_cron.app))
+
+    def test_send_mail_to_admin_on_job_success(self):
+        self.login(self.ADMIN_EMAIL, is_super_admin=True)
+
+        email_subjects = []
+        email_bodies = []
+
+        def _mock_send_mail_to_admin(email_subject, email_body):
+            """Mocks email_manager.send_mail_to_admin() as its not possible to
+            send mail with self.testapp_swap, i.e with the URLs defined in
+            main_cron.
+            """
+            email_subjects.append(email_subject)
+            email_bodies.append(email_body)
+
+        send_mail_to_admin_swap = self.swap(
+            email_manager, 'send_mail_to_admin', _mock_send_mail_to_admin)
+
+        with self.testapp_swap, send_mail_to_admin_swap:
+            self.get_html_response('/cron/mail/admin/job_status')
+
+        self.assertEqual(email_subjects, ['MapReduce status report'])
+        self.assertEqual(email_bodies, ['All MapReduce jobs are running fine.'])
+
+        self.logout()
+
+    def test_send_mail_to_admin_on_job_failure(self):
+        self.login(self.ADMIN_EMAIL, is_super_admin=True)
+
+        email_subjects = []
+        email_bodies = []
+
+        def _mock_send_mail_to_admin(email_subject, email_body):
+            """Mocks email_manager.send_mail_to_admin() as its not possible to
+            send mail with self.testapp_swap, i.e with the URLs defined in
+            main_cron.
+            """
+            email_subjects.append(email_subject)
+            email_bodies.append(email_body)
+
+        send_mail_to_admin_swap = self.swap(
+            email_manager, 'send_mail_to_admin', _mock_send_mail_to_admin)
+
+        job_id = SampleMapReduceJobManager.create_new()
+        SampleMapReduceJobManager.enqueue(
+            job_id, taskqueue_services.QUEUE_NAME_DEFAULT)
+        self.assertEqual(
+            self.count_jobs_in_taskqueue(
+                taskqueue_services.QUEUE_NAME_DEFAULT), 1)
+
+        self.process_and_flush_pending_tasks()
+        self.assertEqual(
+            SampleMapReduceJobManager.get_status_code(job_id),
+            jobs.STATUS_CODE_COMPLETED)
+
+        shard_state_model_class = mapreduce_model.ShardState
```

Comment:
Add some line comments to explain the context of what you are doing here.

You can also drop the newline below this line -- I think lines 111-116 are conceptually one "block".

Reply:
done